core/test_utils.py

```diff
@@ -0,0 +1,65 @@
+import unittest
+from unittest import mock
+import numpy as np
+from core import utils
+
+
+class TestUtils(unittest.TestCase):
+    @mock.patch("core.utils.np.random.normal", return_value=np.array([1, 0]))
+    def test_randomly_perturb_direction(self, mock_normal: mock.Mock):
+        new_direction = utils.randomly_perturb_direction(np.array([0, 1]), 1)
+        expected_direction = np.array([np.sqrt(0.5), np.sqrt(0.5)])
```

Comment:
I think it would be more human-readable to replace sqrt(0.5) with the approx. value of 0.70711. You can put in a comment that this is sqrt(0.5) if that seems useful.